Index: xorg-server-1.3.0.0.dfsg/hw/xfree86/parser/Makefile.am
===================================================================
--- xorg-server-1.3.0.0.dfsg.orig/hw/xfree86/parser/Makefile.am	2007-02-16 16:03:12.000000000 +0000
+++ xorg-server-1.3.0.0.dfsg/hw/xfree86/parser/Makefile.am	2007-05-26 18:17:17.000000000 +0000
@@ -26,7 +26,7 @@
 	DRI.c \
 	Extensions.c
 
-AM_CFLAGS = $(DIX_CFLAGS) $(XORG_CFLAGS)
+AM_CFLAGS = $(DIX_CFLAGS) $(XORG_CFLAGS) -fPIC
 
 EXTRA_DIST = \
 	Configint.h \
Index: xorg-server-1.3.0.0.dfsg/hw/xfree86/parser/Makefile.in
===================================================================
--- xorg-server-1.3.0.0.dfsg.orig/hw/xfree86/parser/Makefile.in	2007-05-26 18:11:37.000000000 +0000
+++ xorg-server-1.3.0.0.dfsg/hw/xfree86/parser/Makefile.in	2007-05-26 18:22:15.000000000 +0000
@@ -577,7 +375,7 @@
 	DRI.c \
 	Extensions.c
 
-AM_CFLAGS = $(DIX_CFLAGS) $(XORG_CFLAGS)
+AM_CFLAGS = $(DIX_CFLAGS) $(XORG_CFLAGS) -fPIC
 EXTRA_DIST = \
 	Configint.h \
 	configProcs.h \
