---
 hw/xfree86/common/xf86AutoConfig.c |   32 ++++++++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

--- a/hw/xfree86/common/xf86AutoConfig.c
+++ b/hw/xfree86/common/xf86AutoConfig.c
@@ -44,6 +44,10 @@
 #endif
 #include "dirent.h"
 
+#if defined(__arm__) && defined(__linux__)
+# include "loaderProcs.h"
+#endif
+
 #ifdef sun
 # include <sys/visual_io.h>
 # include <ctype.h>
@@ -423,6 +427,28 @@ matchDriverFromFiles (char** matches, ui
 }
 #endif /* __linux__ */
 
+#if defined(__arm__) && defined(__linux__)
+static int
+test_sysfs_device (char * device_name, char * driver_name)
+{
+    DIR* dir = opendir("/sys/devices/platform");
+    struct dirent *current_dir;
+    int len = strlen(device_name);
+
+    while (current_dir = readdir(dir)) {
+        if (strlen(current_dir->d_name) >= len && strncmp(device_name, current_dir->d_name, len) == 0)
+            break;
+    }
+    closedir(dir);
+    if(!current_dir)
+        return 0;
+
+    if (!LoadModule(driver_name, NULL, NULL, NULL, NULL, NULL, NULL, NULL))
+        return 0;
+    return 1;
+}
+#endif /* defined(__arm__) && defined(__linux__) */
+
 static void
 listPossibleVideoDrivers(char *matches[], int nmatches)
 {
@@ -534,7 +560,18 @@ listPossibleVideoDrivers(char *matches[]
 #if !defined(__linux__) && defined(__sparc__)
 	matches[i++] = xnfstrdup("wsfb");
 #else
+#if defined(__linux__) && defined(__arm__)
+        if (test_sysfs_device("mxc_gpu", "imx"))
+           matches[i++] = xnfstrdup("imx");
+        else if (test_sysfs_device("dovefb", "dovefb"))
+           matches[i++] = xnfstrdup("dovefb");
+        else if (test_sysfs_device("omapfb", "omapfb"))
+           matches[i++] = xnfstrdup("omapfb");
+        else
+	    matches[i++] = xnfstrdup("fbdev");
+#else
 	matches[i++] = xnfstrdup("fbdev");
+#endif /* defined(__linux__) && defined(__arm__) */
 #endif
     }
 }
