Index: xorg-server/hw/xfree86/common/xf86RandR.c
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86RandR.c
+++ xorg-server/hw/xfree86/common/xf86RandR.c
@@ -54,6 +54,8 @@
 {
     if (mode->VRefresh)
 	return (int) (mode->VRefresh + 0.5);
+    else if (mode->Clock == 0)
+	return 0;
     else
 	return (int) (mode->Clock * 1000.0 / mode->HTotal / mode->VTotal + 0.5);
 }
