diff -Nurp xorg-server-1.4.99.906-patched2/hw/xfree86/os-support/linux/lnx_acpi.c xorg-server-1.4.99.906-working2/hw/xfree86/os-support/linux/lnx_acpi.c
--- xorg-server-1.4.99.906-patched2/hw/xfree86/os-support/linux/lnx_acpi.c	2008-09-01 13:32:09.000000000 -0700
+++ xorg-server-1.4.99.906-working2/hw/xfree86/os-support/linux/lnx_acpi.c	2008-09-01 14:06:56.000000000 -0700
@@ -133,6 +133,10 @@ lnxACPIOpen(void)
     int fd;    
     struct sockaddr_un addr;
     int r = -1;
+    unsigned int acpi_warning_msg_interval = 10; /* interval between error prints */
+    static unsigned int acpi_warning_msg_timer = 0;
+    unsigned int acpi_warning_max_count = 20; /* Don't print error more than this many times total */
+    static unsigned int acpi_warning_count = 0;
 
 #ifdef DEBUG
     ErrorF("ACPI: OSPMOpen called\n");
@@ -148,8 +152,22 @@ lnxACPIOpen(void)
 	addr.sun_family = AF_UNIX;
 	strcpy(addr.sun_path, ACPI_SOCKET);
 	if ((r = connect(fd, (struct sockaddr*)&addr, sizeof(addr))) == -1) {
-	    xf86MsgVerb(X_WARNING,3,"Open ACPI failed (%s) (%s)\n", ACPI_SOCKET,
-	    	strerror(errno));
+            if (acpi_warning_count < acpi_warning_max_count) {
+                if (acpi_warning_msg_timer == 0) {
+                    /* Don't fill log with warning (LP: #247195) */
+                    xf86MsgVerb(X_WARNING,3,"Open ACPI failed (%s) (%s)\n", ACPI_SOCKET,
+                                strerror(errno));
+                    acpi_warning_msg_timer = acpi_warning_msg_interval;
+                    acpi_warning_count++;
+
+                    if (acpi_warning_count == acpi_warning_max_count) {
+                        xf86MsgVerb(X_WARNING,1,"Has the acpid daemon died?  Try restarting it.  Suppressing further ACPI failure messages. (see LP: #247195)\n");
+                    }
+
+                } else {
+                    --acpi_warning_msg_timer;
+                }
+            }
 	    shutdown(fd, 2);
 	    close(fd);
 	    return NULL;
