--- a/mi/mipointer.c	2009-02-04 19:35:31.000000000 -0500
+++ b/mi/mipointer.c	2009-02-04 19:38:21.000000000 -0500
@@ -332,7 +332,15 @@
 	pPointer->pScreen = pScreen;
     }
 
-    if (changedScreen)
+    /* Don't call USFS if we use Xinerama, otherwise the root window is
+     * updated to the second screen, and we never receive any events.
+     * (FDO bug #18668) */
+
+    if (changedScreen 
+#ifdef PANORAMIX
+		&& noPanoramiXExtension
+#endif
+	)
         UpdateSpriteForScreen (pDev, pScreen) ;
 }
 
