From ce50bfd3369686cfecee5a138bd84ef1107a249d Mon Sep 17 00:00:00 2001
From: Michel Dänzer <michel@tungstengraphics.com>
Date: Thu, 18  Oct  2007  15:44:48  +0000
Subject: EXA: Skip empty glyphs.

---
diff --git a/exa/exa_render.c b/exa/exa_render.c
index 65e67d8..cc2f59d 100644
--- a/exa/exa_render.c
+++ b/exa/exa_render.c
@@ -1246,6 +1246,7 @@ exaGlyphs (CARD8	op,
 	    y1 = y - glyph->info.y;
 
 	    if (x1 >= pCmpDrw->width || y1 >= pCmpDrw->height ||
+		glyph->info.width == 0 || glyph->info.height == 0 ||
 		(x1 + glyph->info.width) <= 0 || (y1 + glyph->info.height) <= 0)
 		goto nextglyph;
 
--
cgit v0.7.2-37-g538c
