diff -Nurp patched/mi/mieq.c working/mi/mieq.c
--- patched/mi/mieq.c	2009-01-15 23:18:21.000000000 -0800
+++ working/mi/mieq.c	2009-01-15 23:23:01.000000000 -0800
@@ -354,17 +354,15 @@ mieqProcessInputEvents(void)
             NewCurrentScreen (dev, DequeueScreen(dev), x, y);
         }
         else {
-            /* Force a copy of the key class into the VCK so that the layout
-               is transferred. */
-            if (event->u.u.type == DeviceKeyPress ||
-                event->u.u.type == DeviceKeyRelease)
-                CopyKeyClass(dev, master);
-
-            if (master)
-                CopyGetMasterEvent(master, event,
-                                   &master_event, nevents);
-            else
-                master_event = NULL;
+            if (master) {
+                /* Force a copy of the key class into the VCK so that the layout
+                   is transferred. */
+                if (event->u.u.type == DeviceKeyPress ||
+                    event->u.u.type == DeviceKeyRelease)
+                    CopyKeyClass(dev, master);
+
+                CopyGetMasterEvent(master, event, masterEvents, nevents);
+            }
 
             /* If someone's registered a custom event handler, let them
              * steal it. */
