diff --git a/mi/mipointer.c b/mi/mipointer.c
index de6698a..907f2e4 100644
--- a/mi/mipointer.c
+++ b/mi/mipointer.c
@@ -362,6 +362,10 @@ miPointerWarpCursor(DeviceIntPtr pDev, ScreenPtr pScreen, int x, int y)
     SetupScreen(pScreen);
     pPointer = MIPOINTER(pDev);
 
+    /* Null pointer causes crash on keyrepeat with Xinerama LP: (#324465) */
+    if (pPointer == NULL)
+        return;
+
     if (pPointer->pScreen != pScreen) {
         (*pScreenPriv->screenFuncs->NewEventScreen) (pDev, pScreen, TRUE);
         changedScreen = TRUE;
