diff --git a/exa/exa_migration.c b/exa/exa_migration.c
index 571650c..733379f 100644
--- a/exa/exa_migration.c
+++ b/exa/exa_migration.c
@@ -166,7 +166,7 @@ exaCopyDirty(ExaMigrationPtr migrate, RegionPtr pValidDst, RegionPtr pValidSrc,
 		static Bool firsttime = TRUE;
 
 		if (firsttime) {
-		    ErrorF("%s: Pending damage region empty!\n", __func__);
+		    /* ErrorF("%s: Pending damage region empty!\n", __func__); */
 		    firsttime = FALSE;
 		}
 	    }
