=== modified file 'hw/vfb/InitOutput.c'
Index: xorg-server/hw/vfb/InitOutput.c
===================================================================
--- xorg-server.orig/hw/vfb/InitOutput.c	2011-01-10 16:10:26.000000000 -0600
+++ xorg-server/hw/vfb/InitOutput.c	2011-01-10 17:02:37.000000000 -0600
@@ -66,6 +66,7 @@
 #endif /* HAS_SHM */
 #include "dix.h"
 #include "miline.h"
+#include "randrstr.h"
 
 #define VFB_DEFAULT_WIDTH      1280
 #define VFB_DEFAULT_HEIGHT     1024
@@ -881,6 +882,12 @@
 
     if (!ret) return FALSE;
 
+    if (!miRandRInit(pScreen)) {
+	xf86DrvMsg(pScreen->index, X_WARNING,
+		   "Could not initialise RANDR\n");
+	return FALSE;
+    }
+
     pScreen->InstallColormap = vfbInstallColormap;
     pScreen->UninstallColormap = vfbUninstallColormap;
     pScreen->ListInstalledColormaps = vfbListInstalledColormaps;
