=== modified file 'hw/vfb/InitOutput.c'
--- a/hw/vfb/InitOutput.c
+++ b/hw/vfb/InitOutput.c
@@ -918,6 +918,12 @@ vfbScreenInit(int index, ScreenPtr pScre
 	fbPictureInit (pScreen, 0, 0);
 #endif
 
+    if (!miRandRInit(pScreen))
+        /* Not sure how to emit warnings in xorg. Should warn here though:
+	 * ("Could not initialise RANDR\n");
+	 */
+        return FALSE;
+
     miInitializeBackingStore(pScreen);
 
     /*
