diff --git a/randr/randr.c b/randr/randr.c
index 9f3df5f..52443b5 100644
--- a/randr/randr.c
+++ b/randr/randr.c
@@ -233,7 +233,7 @@ RRScreenInit(ScreenPtr pScreen)
     /*
      * Calling function best set these function vectors
      */
-    pScrPriv->rrGetInfo = 0;
+    pScrPriv->rrGetInfo = NULL;
     pScrPriv->maxWidth = pScrPriv->minWidth = pScreen->width;
     pScrPriv->maxHeight = pScrPriv->minHeight = pScreen->height;
 
