diff --git a/os/utils.c b/os/utils.c
index 30592d2..e9c17d0 100644
--- a/os/utils.c
+++ b/os/utils.c
@@ -506,6 +506,8 @@ UseMsg(void)
     ErrorF("-nolisten string       don't listen on protocol\n");
     ErrorF("-noreset               don't reset after last client exists\n");
     ErrorF("-background [none]     create root window with no background\n");
+    ErrorF
+        ("-nr                    (Ubuntu-specific) Synonym for -background none\n");
     ErrorF("-reset                 reset after last client exists\n");
     ErrorF("-p #                   screen-saver pattern duration (minutes)\n");
     ErrorF("-pn                    accept failure to listen on all ports\n");
@@ -825,6 +827,8 @@ ProcessCommandLine(int argc, char *argv[])
                     UseMsg();
             }
         }
+        else if (strcmp(argv[i], "-nr") == 0)
+            bgNoneRoot = TRUE;
         else if (strcmp(argv[i], "-maxbigreqsize") == 0) {
             if (++i < argc) {
                 long reqSizeArg = atol(argv[i]);
