From b96275c4cdb164aa71f7aa9fbf88be18886d1936 Mon Sep 17 00:00:00 2001
From: Benjamin Otte <otte@redhat.com>
Date: Tue, 7 Feb 2012 18:01:25 +1000
Subject: [PATCH] dix: fix an out-of-memory crash

Reviewed-by: Peter Hutterer <peter.hutterer@who-t.net>
Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
---
 dix/inpututils.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/dix/inpututils.c b/dix/inpututils.c
index d279c1d..9e0c551 100644
--- a/dix/inpututils.c
+++ b/dix/inpututils.c
@@ -432,6 +432,9 @@ valuator_mask_new(int num_valuators)
      * flying-car future, when we can dynamically alloc the masks and are
      * not constrained by signals, we can start using num_valuators */
     ValuatorMask *mask = calloc(1, sizeof(ValuatorMask));
+    if (mask == NULL)
+        return NULL;
+
     mask->last_bit = -1;
     return mask;
 }
-- 
1.7.8.3

