Index: xorg-server/hw/xfree86/doc/man/xorg.conf.man.pre
===================================================================
--- xorg-server.orig/hw/xfree86/doc/man/xorg.conf.man.pre	2008-06-18 10:28:16.000000000 +0300
+++ xorg-server/hw/xfree86/doc/man/xorg.conf.man.pre	2008-06-18 12:06:37.000000000 +0300
@@ -356,11 +356,11 @@
 .PP
 .RS 4
 .nf
-.I __projectroot__/lib/X11/fonts/misc/
-.I __projectroot__/lib/X11/fonts/TTF/
-.I __projectroot__/lib/X11/fonts/Type1/
-.I __projectroot__/lib/X11/fonts/75dpi/
-.I __projectroot__/lib/X11/fonts/100dpi/
+.I __projectroot__/share/X11/fonts/misc/
+.I __projectroot__/share/X11/fonts/TTF/
+.I __projectroot__/share/X11/fonts/Type1/
+.I __projectroot__/share/X11/fonts/75dpi/
+.I __projectroot__/share/X11/fonts/100dpi/
 .fi
 .RE
 .PP
@@ -368,13 +368,13 @@
 .PP
 .RS 4
 .nf
-.I __projectroot__/lib/X11/fonts/local/
-.I __projectroot__/lib/X11/fonts/misc/
-.I __projectroot__/lib/X11/fonts/75dpi/:unscaled
-.I __projectroot__/lib/X11/fonts/100dpi/:unscaled
-.I __projectroot__/lib/X11/fonts/Type1/
-.I __projectroot__/lib/X11/fonts/75dpi/
-.I __projectroot__/lib/X11/fonts/100dpi/
+.I __projectroot__/share/X11/fonts/local/
+.I __projectroot__/share/X11/fonts/misc/
+.I __projectroot__/share/X11/fonts/75dpi/:unscaled
+.I __projectroot__/share/X11/fonts/100dpi/:unscaled
+.I __projectroot__/share/X11/fonts/Type1/
+.I __projectroot__/share/X11/fonts/75dpi/
+.I __projectroot__/share/X11/fonts/100dpi/
 .fi
 .RE
 .PP
