
diff -Nurp patched/hw/xfree86/common/xf86AutoConfig.c working/hw/xfree86/common/xf86AutoConfig.c
--- patched/hw/xfree86/common/xf86AutoConfig.c	2008-10-17 19:48:15.000000000 -0700
+++ working/hw/xfree86/common/xf86AutoConfig.c	2008-10-17 19:43:19.000000000 -0700
@@ -422,6 +422,20 @@ matchDriverFromFiles (char** matches, ui
         }
         direntry = readdir(idsdir);
     }
+    /* If we failed to find any driver, at least try 'vesa' (LP: #261977) */
+    if (matches[0] == NULL) {
+        xf86Msg(X_INFO, "No matches found for this device in %s\n",  PCI_TXT_IDS_PATH);
+
+        matches[0] = (char*)xalloc(sizeof(char) * 5);
+        if (!matches[0]) {
+            xf86Msg(X_ERROR, "Could not allocate space for the module name. Exiting.\n");
+            goto end;
+        }
+
+        xf86Msg(X_DEFAULT, "Registering 'vesa' as fallback\n");
+        strncpy(matches[0], "vesa", 5);
+    }
+
  end:
     xfree(line);
     closedir(idsdir);
