diff --git a/hw/dmx/dmx.h b/hw/dmx/dmx.h
index 05e5fab..1a71417 100644
--- a/hw/dmx/dmx.h
+++ b/hw/dmx/dmx.h
@@ -341,14 +341,14 @@ do {									\
 #define _MAXSCREENSALLOCF(o,size,fatal)                                 \
     do {                                                                \
         if (!o) {                                                       \
-            o = xcalloc((size), sizeof(*(o)));                          \
+            o = calloc((size), sizeof(*(o)));                          \
             if (!o && fatal) FatalError(MAXSCREEN_FAILED_TXT #o);       \
         }                                                               \
     } while (0)
 #define _MAXSCREENSALLOCR(o,size,retval)                                \
     do {                                                                \
         if (!o) {                                                       \
-            o = xcalloc((size), sizeof(*(o)));                          \
+            o = calloc((size), sizeof(*(o)));                          \
             if (!o) return retval;                                      \
         }                                                               \
     } while (0)
diff --git a/hw/dmx/dmxextension.c b/hw/dmx/dmxextension.c
index d20c844..6fd02dc 100644
--- a/hw/dmx/dmxextension.c
+++ b/hw/dmx/dmxextension.c
@@ -1121,7 +1121,7 @@ static void dmxBERestoreRenderGlyph(pointer value, XID id, pointer n)
     }
 
     /* Now allocate the memory we need */
-    images = xcalloc(len_images, sizeof(char));
+    images = calloc(len_images, sizeof(char));
     gids   = xalloc(glyphSet->hash.tableEntries*sizeof(Glyph));
     glyphs = xalloc(glyphSet->hash.tableEntries*sizeof(XGlyphInfo));
 
diff --git a/hw/dmx/glxProxy/glxutil.c b/hw/dmx/glxProxy/glxutil.c
index 080992e..ab1edef 100644
--- a/hw/dmx/glxProxy/glxutil.c
+++ b/hw/dmx/glxProxy/glxutil.c
@@ -69,7 +69,7 @@ __glXCalloc(size_t numElements, size_t elementSize)
     if ((numElements == 0) || (elementSize == 0)) {
 	return NULL;
     }
-    addr = xcalloc(numElements, elementSize);
+    addr = calloc(numElements, elementSize);
     if (addr == NULL) {
 	/* XXX: handle out of memory error */
 	return NULL;
