From d502521c3669f3f22b94c39a64ab63bfd92c6a97 Mon Sep 17 00:00:00 2001
From: Pierre Willenbrock <pierre@pirsoft.dnsalias.org>
Date: Tue, 23  Oct  2007  14:45:13  +0000
Subject: EXA: Fix off-by-one in polyline drawing.

---
diff --git a/exa/exa_accel.c b/exa/exa_accel.c
index 52cc5c4..5fb72d7 100644
--- a/exa/exa_accel.c
+++ b/exa/exa_accel.c
@@ -658,7 +658,7 @@ exaPolylines(DrawablePtr pDrawable, GCPtr pGC, int mode, int npt,
     x1 = ppt[0].x;
     y1 = ppt[0].y;
     /* If we have any non-horizontal/vertical, fall back. */
-    for (i = 0; i < npt; i++) {
+    for (i = 0; i < npt - 1; i++) {
 	if (mode == CoordModePrevious) {
 	    x2 = x1 + ppt[i + 1].x;
 	    y2 = y1 + ppt[i + 1].y;
--
cgit v0.7.2-37-g538c
