Index: xorg-server/hw/xfree86/common/xf86AutoConfig.c
===================================================================
--- xorg-server.orig/hw/xfree86/common/xf86AutoConfig.c	2008-03-31 09:57:50.000000000 +0300
+++ xorg-server/hw/xfree86/common/xf86AutoConfig.c	2008-03-31 09:57:57.000000000 +0300
@@ -180,7 +180,7 @@
                 case 0x8108:
                     return "psb";
                 default:
-                    return "i810";
+                    return "intel";
             }
 	case 0x102b:		    return "mga";
 	case 0x10c8:		    return "neomagic";
