--- xorg-server-1.4.1~git20080131/hw/xfree86/common/xf86AutoConfig.c.orig	2008-02-01 05:31:28.000000000 +0200
+++ xorg-server-1.4.1~git20080131/hw/xfree86/common/xf86AutoConfig.c	2008-04-19 20:05:41.000000000 +0300
@@ -160,12 +160,13 @@ videoPtrToDriverName(pciVideoPtr info)
 {
     /*
      * things not handled yet:
-     * amd/cyrix/nsc
+     * cyrix/nsc. should be merged into geode anyway.
      * xgi
      */
 
     switch (info->vendor)
     {
+	case 0x1022:                return "geode";
 	case 0x1142:		    return "apm";
 	case 0xedd8:		    return "ark";
 	case 0x1a03:		    return "ast";
