--- xorg-server.orig/hw/xfree86/common/xf86AutoConfig.c	2008-05-10 16:33:59.000000000 +0100
+++ xorg-server/hw/xfree86/common/xf86AutoConfig.c	2008-05-10 16:47:34.000000000 +0100
@@ -372,7 +372,7 @@
 
     if (!info) {
         ErrorF("Could not get primary PCI info\n");
-        goto end;
+        goto nopci;
     }
 
     idsdir = opendir("/usr/share/xserver-xorg/pci");
@@ -453,6 +453,13 @@
         closedir(idsdir);
     }
 
+    /*
+     * We jump here if there's no PCI info. We may be on a platform like the
+     * PS3 where we should fallback to the platform default driver like fbdev
+     * or vesa.
+     */
+    nopci:
+
     /* TODO Handle multiple drivers claiming to support the same PCI ID */
     if (matches[0]) {
         chosen_driver = matches[0];
