diff --git a/glx/Makefile.am b/glx/Makefile.am
index 591c4ac..5f91946 100644
--- a/glx/Makefile.am
+++ b/glx/Makefile.am
@@ -10,7 +10,7 @@ AM_CFLAGS = \
 	@XLIB_CFLAGS@ \
 	@LIBDRM_CFLAGS@ \
 	@GLX_DEFINES@ \
-	@GLX_ARCH_DEFINES@
+	@GLX_ARCH_DEFINES@ -lglapi
 
 # none yet
 #sdk_HEADERS =
diff --git a/glx/glapi.c b/glx/glapi.c
index ad7329e..8b754e6 100644
--- a/glx/glapi.c
+++ b/glx/glapi.c
@@ -80,13 +80,6 @@ static void init_glapi_relocs(void);
 /*@{*/
 #if defined(GLX_USE_TLS)
 
-PUBLIC TLS struct _glapi_table *_glapi_tls_Dispatch = NULL;
-
-PUBLIC TLS void *_glapi_tls_Context;
-
-PUBLIC const struct _glapi_table *_glapi_Dispatch = NULL;
-PUBLIC const void *_glapi_Context = NULL;
-
 #else
 
 #if defined(THREADS)
